useful_types/experimental.py

```diff
@@ -22,8 +34,33 @@
 # Objects suitable to be passed to sys.settrace, threading.settrace, and similar
 TraceFunction: TypeAlias = Callable[[FrameType, str, Any], Union["TraceFunction", None]]
 
-# Might not work as expected for pyright, see
-#   https://github.com/python/typeshed/pull/9362
-#   https://github.com/microsoft/pyright/issues/4339
-class DataclassInstance(Protocol):
-    __dataclass_fields__: ClassVar[dict[str, Field[Any]]]
+if TYPE_CHECKING:
```

Comment:
I was basing the runtime implementation off my CPython PR https://github.com/python/cpython/pull/102933 -- if we were to add it to CPython eventually, it would be best to add it to `dataclasses.py`, which shouldn't have a dependency on `typing` if we can avoid it.

But it's true that we don't have that same constraint here -- I guess it depends on to what extent we view this package as a "testing ground" for things that we might eventually want to add to the stdlib? If we _are_ looking to maybe add this to CPython one day, it might be good to keep the implementation as close as possible to what it would look like if it were added to CPython

Reply:
I had to delete quite a few of the runtime tests in order to have the same definition at runtime as we have at typingtime. Some of that's for the better (the tests were maybe dubious), but some of them I'm a bit sad about. Nothing too awful though.